packages/gatsby-image/src/index.js

```diff
@@ -69,6 +76,95 @@ function getIO() {
   return io
 }
 
+function generateImageSources(imageVariants) {
+  const initialImage = imageVariants[0]
+
+  if (imageVariants.length === 1) {
```

Comment:
I was thinking that it felt weird to me to spread the source generation logic across two places. I was choosing to combine all source generation logic into the `generateImageSources` function, instead of keeping the implementations separate based on whether there was art direction required or not.

That said, I actually really like the approach you've proposed, and seeing it written out makes it seem much cleaner than what I had. I'll update it to match.

Reply:
> I was thinking that it felt weird to me to spread the source generation logic across two places. 

Yeah, I understand that. It was just not as readable since you had to conditionally handle for 2/4 of those methods for returning a webp source only with the other two methods returning null.

Reply:
The order matters for these `<sources>`, and we also need to render both the Webp and an additional `<source>` for the first image to make sure that art direction works. So it actually needs to look more like this.

```js
{this.state.isVisible && (
  <picture>
    {image.srcSetWebp && (
      <source
        type={`image/webp`}
        media={image.media}
        srcSet={image.srcSetWebp}
        sizes={image.sizes}
      />
    )}
    {image.media && (
      <source
        media={image.media}
        srcSet={image.srcSet}
        sizes={image.sizes}
      />
    )}
    {imageVariants.length > 0 && generateImageSources(imageVariants)}
    <Img
      alt={alt}
      title={title}
      sizes={image.sizes}
      src={image.src}
      crossOrigin={this.props.crossOrigin}
      srcSet={image.srcSet}
      style={imageStyle}
      ref={this.imageRef}
      onLoad={this.handleImageLoaded}
      onError={this.props.onError}
      itemProp={itemProp}
    />
  </picture>
)}
```

Similarly for the base64 / tracedSVG placeholders:

```js
{image.base64 && (
  <picture>
    {image.media && (
      <source media={image.media} srcSet={image.base64} />
    )}
    {imageVariants && generateBase64Sources(imageVariants)}
    <Img src={image.base64} {...placeholderImageProps} />
  </picture>
)}
```

Does that seem acceptable to you? We could also remove the `image.media?` check from these and just render the first source by default (assuming no regressions).

Reply:
@brimtown Yeah, that looks good. React will ignore the media attribute if the prop isn't available, it's just with the current noscript support via string template that it needs to be handled conditionally.

If there were any issues from this approach, a refactor can be done later. The additional source element should be a non-issue. All the duplicated fields and values should compress fine afaik.